File: JavaScript/198.js
/**
 * @param {number[]} nums
 * @return {number}
 */
 var rob = function(nums) {
    if(nums.length === 1)
        return nums[0];
    let dp = [];
    dp[0] = nums[0];
    dp[1] = Math.max(dp[0],nums[1]);
    let  max = dp[1];
    for(let i=2; i<nums.length; ++i) {
        dp[i] = Math.max(dp[i-2]+nums[i],dp[i-1]);
        max = Math.max(dp[i], max);
    }
    return max;
};
// 经典的打家劫舍问题

File: Cpp/198.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if(nums.empty())    return 0;
        int len = nums.size();
        if(len == 1)   return nums[0];
        vector<int> dp = vector<int>(len, 0);
        dp[0] = nums[0];
        dp[1] = max(nums[0], nums[1]);

        for(int i=2; i<len; ++i) {
            dp[i] = max(dp[i-1], dp[i-2]+nums[i]);
        }
        return dp[len-1];
    }
};
// 打家劫舍问题